inginious/frontend/lti_grade_manager.py

```diff
@@ -0,0 +1,153 @@
+        self.database = database
+        self.query_context = (courseid, taskid)
+        self._session_cookie_name = ""  # Disables session scope mechanism in favor of query_context
+        super().__init__(*args, **kwargs)
+
+    def can_set_keys_expiration(self) -> bool:
+        return False  # TODO(mp): I think it's reasonable to clean LTI Launch messages further than a week away tho
+
+    def get_value(self, key: str):
+        entry = self.database.lti_launch.find_one({'key': key, 'context': self.query_context})
+        return entry.get('value') if entry else None
+
+    def set_value(self, key: str, value, exp) -> None:
+        self.database.lti_launch.find_one_and_update({'key': key, 'context': self.query_context},
+                                                     {'$set': {'key': key, 'value': value}}, upsert=True)
+
+    def check_value(self, key: str) -> bool:
+        return bool(self.database.lti_launch.find_one({'key': key, 'context': self.query_context}))
+
+
+class LTIGradeManager(threading.Thread):
+    """ Waits for grading to complete and submit grade to the LTI Platform. """
+    def __init__(self, database, app):
+        super(LTIGradeManager, self).__init__()
+        self.daemon = True
+        self._database = database
+        self._app = app
+        self._queue = queue.Queue()
+        self._stopped = False
+        self._logger = logging.getLogger("inginious.webapp.lti_grade_manager")
+        self.start()
+
+    def stop(self):
+        self._stopped = True
+
+    def run(self):
+        # Load old tasks from the database
+        for todo in self._database.lti_grade_queue.find({}):
+            self._add_to_queue(todo)
+
+        try:
+            while not self._stopped:
+                time.sleep(0.5)
+                data = self._queue.get()
+                mongo_id, username, courseid, taskid, message_launch_id, nb_attempt = data
+
+                try:
+                    course = self._app.course_factory.get_course(courseid)
+                    task = course.get_task(taskid)
+                    grade = self._app.user_manager.get_task_cache(username, courseid, task.get_id())["grade"]
+                except Exception:
+                    self._logger.error("An exception occurred while getting a course/LTI secret/grade in LTIGradeManager.", exc_info=True)
+                    continue
+
+                try:
+                    message_launch = FlaskMessageLaunch.from_cache(message_launch_id, request=None, tool_config=course.lti_tool(), launch_data_storage=MongoLTILaunchDataStorage(self._app.database, courseid, taskid))
+                    launch_data = message_launch.get_launch_data()
+                    ags = message_launch.get_ags()
+
+                    if ags.can_put_grade():
+                        sc = Grade()
+                        # TODO(mp): Is there a better timestamp to set with the score? Submission time? Grading time?
```

Comment:
Maybe it's better to always set the score as the latest for the platform (and therefore keep the `datetime.now()`). This way we ensure the score shown on INGInious is the one that is stored by the platform.